File: Data Structures & Algorithms/islands-and-treasure/submission-0.cs
public class Solution 
{
    public void islandsAndTreasure(int[][] grid) 
    {
        var rows = grid.Length;
        var cols = grid[0].Length;

        for(int r = 0; r < rows; r++)
        {
            for(int c = 0; c < cols; c++)
            {
                if(grid[r][c] == 0)
                {
                    DFS(grid, r, c, 0);
                }
            }
        }
    }

    private void DFS(int[][] grid, int r, int c, int movements)
    {
        if(r < 0 || c < 0 || r >= grid.Length || c >= grid[0].Length || grid[r][c] == -1 || grid[r][c] < movements)
        {
            return;
        }

        grid[r][c] = movements;

        DFS(grid, r + 1, c, movements + 1);
        DFS(grid, r - 1, c, movements + 1);
        DFS(grid, r, c+1, movements + 1);
        DFS(grid, r, c-1, movements + 1);
    }
}

File: Data Structures & Algorithms/max-area-of-island/submission-0.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        rows = len(grid)
        cols = len(grid[0])
        max_area = 0

        visited = set()

        def dfs(r,c):
            if r < 0 or c < 0:
                return 0
            
            if r >= rows or c >= cols:
                return 0 
            
            if (r,c) in visited:
                return 0
            
            if grid[r][c] == 0:
                return 0
            
            visited.add((r,c))
            return 1 + (dfs(r + 1, c) + dfs(r-1, c) + dfs(r, c+1) + dfs(r, c-1))
            
        for r in range(rows):
             for c in range(cols):
                  area = dfs(r,c)
                  max_area = max(max_area, area)

        return max_area
            

